### ISSUE TITLE
Milestone 1 Report - Challenges of Deployment, Dockerizing, and CI/CD processes.

### ISSUE THREAD

**Issue by cakmadam98** [bounswe/bounswe2020group7#184]
A detailed explanation of the challenges that are met during the deployment, dockerizing, and CI/CD processes.
Deadline is 29.11.2020 10:00

**Comment by HUmutOzdemir** [comment]
The technologies that we used in deployment, dockerizing and CI/CD processes are Docker, Docker Compose, Amazon AWS, and Travis-CI. At the begining of the projecti we took an EC2 instance and a MySQL Database instance on Amazon AWS. We used the MySQL database on the Amazon AWS for the development and testing. Whereas the EC2 instanse that we took is our production server.

During our deployment process, we decided to use Docker and Docker Compose. Our aim is gathering frontend, backend and database servers on a single machine on Amazon. So, by using the Dockerfiles that we created, for backend and frontend a Docker image is created if there is any changes in the source codes in a deployment stage. Also, by using the Docker Compose file that we created, we create an artificial network between the frontend, backend and the MySQL Database Docker images. 

By using the approach that I explained in the second paragraph, a deployment script is created on our production server. This script, firstly, pulls the release branch of our git repository. Then, by using the Docker Compose file that we creaed, it creates necessary Docker images and also artificial network between Docker containers. To obtain a CI/CD process, we used a system called Travis-CI. The reason why we choose to use Travis-CI is because it gives a free uage for the public GitHub repositories. By using Travis-CI, we created a pipeline that provides us a mechanism that cıntinously deploys our code to our production server. There are 2 stages of ths pipeline. In the first stage, it runs all the unit tests of the backend by using our test database. Also it concurrently tries to build the frontend code. If there is not any errors on the building phase, it runs all the tests of frontend. This is the end of the first stage. In the case of an existing error during any of the processes in the first stage, it stops its execution and gives an error message to us. 

If there is not any error, the process continues to second stage. Second stage is only available for the pull requests and commits that are done to the "release" brach which is the brach that runs on our production server. In the second stage a machine that is allocated from Travis-CI, connects to our production server using and ssh key. Whenever it connects to our production server, it runs the deployment script of us, which automatically deploys our code to our production server.

During this process whe met with various problems. In the building stage of this CI/CD pipeline, becasue we took an Amazon Linux 2 machine from Amazon, sometimes it got hard to install the necessary tools to our production server. In addition to the Operatin System compatibility issues, the most important issue, which we still can not solve, is the volume problem of the Docker Compose. To protect the data that is stored in our database from deployments, we created a constant volume for our MySQL database. But during the development stage we notice that it did not use the same volume after each deployment, which creates so many errors on our production server. Also another problem that we met is some crash problems during unit tests, initially our CI platform uses the database that is running on our production server. During unit tests which creates some unwanted situations on our database, there can be some errors on our database. As a result of this, we decided to use our Amazon MySQL instance for the tests of CI and tests which are done during development phase.
